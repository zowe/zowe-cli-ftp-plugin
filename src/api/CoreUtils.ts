/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { ICommandProfileTypeConfiguration, IImperativeError, Logger } from "@zowe/imperative";
import * as stream from "stream";
import { TransferMode } from "zos-node-accessor";

export class CoreUtils {

    /**
     * Sleep for a number of milliseconds
     * @param ms - how many miliseconds to sleep or (at least)
     * @returns promise that resolves when the sleep is complete (or after, not guaranteed)
     */
    public static async sleep(ms: number) {
        this.log.debug("Sleeping for %s milliseconds...", ms);
        return new Promise((resolve) => setTimeout(resolve, ms));
    }

    /**
     * Make sure newlines in a string have carriage returns
     * This is required to prevent errors when transferring
     * in ascii mode
     * @param  original - the original string which you would like to ensure has \r before \n
     * @returns a string where every \n has a \r before it
     */
    public static addCarriageReturns(original: string): string {
        this.log.debug("Adding carriage returns to string of length %s", original.length);
        return original.replace(/\r?\n/g, "\r\n");
    }

    /**
     * Read the complete contents of stdin
     * @param stream - A custom stream to read stdin data from
     * @returns contents piped in to stdin
     */
    public static async readStdin(stream?: stream.Readable): Promise<Buffer> {
        return new Promise<Buffer>((resolve, reject) => {
            let stdinContent: Buffer = Buffer.from([]);
            stream = stream || process.stdin;
            stream.resume();

            stream.on("data", (chunk: Buffer) => {
                this.log.trace(`Read data from stdin: ${chunk.toString()}`);
                this.log.debug(`Read ${chunk.length} bytes of data from stdin`);
                stdinContent = Buffer.concat([stdinContent, chunk]);
            });

            stream.once("end", () => {
                this.log.info("Finished reading stdin");
                resolve(stdinContent);
            });

            stream.on("error", (error: Error) => {
                const stdinReadError: IImperativeError = {
                    msg: "Error encountered while reading from stdin",
                    causeErrors: error,
                    additionalDetails: error?.message
                };
                reject(stdinReadError);
            });
        });
    }

    public static async getProfileSchema(): Promise<ICommandProfileTypeConfiguration[]> {
        const ftpProfile = await require("../imperative").profiles as ICommandProfileTypeConfiguration[];
        return ftpProfile;
    }

    protected static get log(): Logger {
        return Logger.getAppLogger();
    }

    public static getBinaryTransferModeOrDefault(isBinary: boolean, rdw = false): TransferMode {
        return (isBinary ?
            (rdw ? TransferMode.BINARY_RDW : TransferMode.BINARY) :
            (rdw ? TransferMode.ASCII_RDW : TransferMode.ASCII)
        );
    }
}
