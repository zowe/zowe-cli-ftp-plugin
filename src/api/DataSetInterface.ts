/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IFTPProgressHandler } from "./IFTPProgressHandler";

/**
 * Bytes per track.
 */
export const TRACK = 56664;

export interface IDownloadDataSetOption {
    /**
     * Optional local file that the dowloaded dataset will be saved to.
     */
    localFile?: string;

    /**
     * Optional repsonse handler for progressing messages.
     */
    progress?: IFTPProgressHandler;

    /**
     * TRANSFER_TYPE_ASCII, TRANSFER_TYPE_BIANRY, TRANSFER_TYPE_ASCII_RDW, or TRANSFER_TYPE_BINARY_RDW defined in CoreUtils.
     */
    transferType?: string;

    /**
    * Optional encoding for download and upload of z/OS data set
    */
    encoding?: string;
}

export interface IUploadDataSetOption {
    /**
     * Optional content that will be uploaded, if `localFile` is not specified.
     */
    content?: Buffer | string;

    /**
     * Data control block (DCB). Refer to https://github.com/IBM/zos-node-accessor/tree/1.0.7#allocate
     */
    dcb?: string;

    /**
     * Optional local file that will be uploaded.
     */
    localFile?: string;

    /**
     * TRANSFER_TYPE_ASCII or TRANSFER_TYPE_BIANRY defined in CoreUtils.
     */
    transferType?: string;

    /**
    * Optional encoding for download and upload of z/OS data set
    */
    encoding?: string;
}

export interface IAllocateDataSetOption {

    /**
     * Data control block (DCB). Refer to https://github.com/IBM/zos-node-accessor/tree/1.0.7#allocate
     */
    dcb?: string;
}

/**
 * Copy Dataset Options
 */
export interface ICopyDataSetOptions {
    /**
     * Required: Name of the SOURCE dataset
     */
    fromDsn: string;

    /**
     * Required: Name of the TARGET dataset
     */
    toDsn: string;

    /**
     * Optional: Progress indicator object/task
     */
    progress?: IFTPProgressHandler;

    /**
     * Optional: Indicator to force a replacement
     */
    replace?: boolean;
}

/**
 * Detailed allocation options for a dataset
 */
export interface IDataSetDetailedAllocationOptions {
    /**
     * Dataset volume serial
     */
    volume?: string; // Not supported by connection.allocateDataset

    /**
     * Dataset record format
     */
    recfm?: string;

    /**
     * Dataset block size
     */
    BLOCKSIze?: string; // Strange mapping

    /**
     * Dataset record length
     */
    lrecl?: string;

    /**
     * Dataset organization (PS vs PO)
     */
    dsorg?: string;

    /**
     * The following values are not supported by the allocateDataset method of zos-node-accessor@v1
     * However, they are return as allocation details for a dataset
     */
    // BLocks: ds.??? // Strage mapping + Not returned by connection.listDataset
    // CYlinders: ds.??? // Strage mapping + Not returned by connection.listDataset
    // TRacks: ds.??? // Strage mapping + Not returned by connection.listDataset
    // Directory: ds.??? // Strage mapping + Not returned by connection.listDataset
    // PRImary: ds.??? // Strage mapping + Not returned by connection.listDataset
    // SECondary: ds.??? // Strage mapping + Not returned by connection.listDataset
}

// When DataSetUtilsV2 for zos-node-accessor v2 is ready, alias DataSetUtilsV2 to DataSetUtils.
// export { DataSetUtils as DataSetUtils };
