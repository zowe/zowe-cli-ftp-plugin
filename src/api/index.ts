/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

export * as zosNodeAccessor from "zos-node-accessor";
export { TransferMode } from "zos-node-accessor";

export * from "./doc";

export * from "./CoreUtils";
export * from "./FTPConfig";
export * from "./JobUtils";
export * from "./DataSetUtils";
export * from "./UssUtils";
