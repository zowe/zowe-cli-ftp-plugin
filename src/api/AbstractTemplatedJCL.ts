/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IO, Logger, TextUtils } from "@zowe/imperative";
import { CoreUtils } from "./CoreUtils";

export abstract class AbstractTemplatedJCL {

    /**
     * Default job template for submitting your JCL
     */
    public abstract readonly DEFAULT_TEMPLATE: string;

    /**
     *
     * @param jobCardFile - file containing a job card which will be prepended to the job
     * @param  substitutionValues -
     * @param  templateFile - override the default JCL template with a file
     */
    public getJcl(jobCardFile: string, substitutionValues: {[key: string]: string}, templateFile?: string) {
        let jcl = IO.readFileSync(jobCardFile).toString();
        if (!/\r?\nZ/.test(jcl)) {
            // add a newline to the end of the jcl if there is none
            jcl += "\r\n";
        }
        if (templateFile != null) {
            // we
            jcl += IO.readFileSync(templateFile).toString();
        } else {
            jcl += this.DEFAULT_TEMPLATE;
        }
        return TextUtils.renderWithMustache(CoreUtils.addCarriageReturns(jcl), substitutionValues);
    }

    protected get log(): Logger {
        return Logger.getAppLogger();
    }
}
