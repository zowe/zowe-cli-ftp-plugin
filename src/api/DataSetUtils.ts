/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import * as fs from "fs";

import { ImperativeError, IO, Logger } from "@zowe/imperative";
import { CoreUtils, TRANSFER_TYPE_ASCII, TRANSFER_TYPE_ASCII_RDW, TRANSFER_TYPE_BINARY, TRANSFER_TYPE_BINARY_RDW } from "./CoreUtils";
import { IAllocateDataSetOption, ICopyDataSetOptions, IDataSetDetailedAllocationOptions, IDownloadDataSetOption, IUploadDataSetOption, TRACK } from "./DataSetInterface";
import { StreamUtils } from "./StreamUtils";
import { IFTPProgressHandler } from "./IFTPProgressHandler";

export class DataSetUtils {

    /**
     * Lists datasets with the specified dataset name pattern.
     *
     * @param connection - zos-node-accessor connection
     * @param pattern - dataset name pattern
     * @returns dataset entries
     */
    public static async listDataSets(connection: any, pattern: string): Promise<any[]> {
        this.log.debug("Listing data sets matching pattern '%s' via FTP", pattern);
        const files = await connection.listDataset(pattern);

        this.log.debug("Found %d matching data sets", files.length);
        const filteredFiles = files.map((file: any) => CoreUtils.addLowerCaseKeysToObject(file));
        return filteredFiles;
    }

    /**
     * Lists members of PDS dataset of the specified dataset name.
     *
     * @param connection - zos-node-accessor connection
     * @param dsn - fully-qualified dataset name without quotes
     * @returns member entries
     */
    public static async listMembers(connection: any, dsn: string): Promise<any[]> {
        this.log.debug("List members of %s", dsn);

        const datasetname = dsn + "(*)";
        const members = await connection.listDataset(datasetname);

        this.log.debug("Found %d members", members.length);
        const filteredMembers = members.map((file: any) => CoreUtils.addLowerCaseKeysToObject(file));

        return filteredMembers;
    }

    /**
     * Delets the dataset of the specified dataset name.
     *
     * @param connection - zos-node-accessor connection
     * @param dsn - fully-qualified dataset name without quotes
     */
    public static async deleteDataSet(connection: any, dsn: string): Promise<void> {
        this.log.debug("Deleting data set '%s'", dsn);
        await connection.deleteDataset("'" + dsn + "'");
    }

    /**
     * Renames the dataset.
     *
     * @param connection - zos-node-accessor connection
     * @param oldDsn - old fully-qualified dataset name without quotes
     * @param newDsn - new fully-qualified dataset name without quotes
     */
    public static async renameDataSet(connection: any, oldDsn: string, newDsn: string): Promise<void> {
        this.log.debug("Attempting to rename data set from '%s' to '%s'", oldDsn, newDsn);
        await connection.rename(oldDsn, newDsn);
    }

    /**
     * Downloads the dataset. The contents will saved in local file, if localFile is provided in option. Otherwise, the contents
     * will be returned in Buffer.
     *
     * @param connection - zos-node-accessor connection
     * @param dsn - fully-qualified dataset name without quotes
     * @param option - download option
     * @returns dataset contents in Buffer, if localFile is not provided in option. Otherwise, undefined will be returned.
     */
    public static async downloadDataSet(connection: any, dsn: string, option: IDownloadDataSetOption): Promise<Buffer> {
        const files = await connection.listDataset(dsn) ?? [];
        if (files.length === 0) {
            throw new ImperativeError({msg: `The dataset "${dsn}" doesn't exist.`});
        }

        const estimatedSize = parseInt(files[0].Used, 10) * TRACK;
        let encoding;
        if (option.encoding) {
            encoding= "sbd=(" + option.encoding + ",ISO8859-1)";
        }
        const transferType = option.transferType || TRANSFER_TYPE_ASCII;

        let buffer;
        let length;
        const stream = await connection.getDataset(dsn, transferType, true, encoding);
        if (option.localFile) {
            this.log.debug("Downloading data set '%s' to local file '%s' in transfer mode '%s'",
                dsn, option.localFile, transferType);
            IO.createDirsSyncFromFilePath(option.localFile);
            const writable = fs.createWriteStream(option.localFile);
            length = await StreamUtils.streamToStream(estimatedSize, stream, writable, option.progress);
        } else {
            this.log.debug("Downloading data set '%s' in transfer mode '%s'", dsn, transferType);
            buffer = await StreamUtils.streamToBuffer(estimatedSize, stream, option.progress);
            length = buffer.byteLength;
        }
        this.log.info("Successfully downloaded %d bytes of content from %s", length, dsn);
        return Promise.resolve(buffer);
    }

    /**
     * Uploads the content to the specified dataset.
     *
     * @param connection - zos-node-accessor connection
     * @param dsn - fully-qualified dataset name without quotes
     * @param option - upload option
     */
    public static async uploadDataSet(connection: any, dsn: string, option: IUploadDataSetOption): Promise<void> {
        const transferType = option.transferType || TRANSFER_TYPE_ASCII;
        let content = option.content;
        let siteparm;
        let encoding;
        if (option.encoding) {
            encoding= "sbd=(" + option.encoding + ",ISO8859-1)";
            siteparm = option.dcb + " " + encoding;
        } else {
            siteparm = option.dcb;
        }
        if (option.localFile) {
            this.log.debug("Attempting to upload from local file '%s' to data set '%s' in transfer mode '%s'",
                option.localFile, dsn, transferType);
            content = IO.readFileSync(option.localFile, undefined, transferType === TRANSFER_TYPE_BINARY);
        } else {
            this.log.debug("Attempting to upload to data set '%s' in transfer mode '%s'", dsn, option.transferType);
        }
        if (transferType === TRANSFER_TYPE_ASCII) {
            content = CoreUtils.addCarriageReturns(content.toString());
        }
        await connection.uploadDataset(content, "'" + dsn + "'", transferType, siteparm);
    }

    /**
     * Map the allocation details from a dataset returned from the zos-node-accessor package to IDs that it accepts for allocation purposes
     * @param ds Descriptor for dataset allocation options
     * @returns Object with mapping required for allocation purposes (BLOGKSIze vs blksz)
     */
    public static mapAllocationOptions(ds: any): IDataSetDetailedAllocationOptions {
        // supported options: https://github.com/IBM/zos-node-accessor/blob/1.0.x/lib/zosAccessor.js#LL122C68-L122C68
        return {
            volume: ds.volume, // Not supported by connection.allocateDataset
            recfm: ds.recfm,
            BLOCKSIze: ds.blksz, // Strange mapping
            lrecl: ds.lrecl,
            dsorg: ds.dsorg,
            // BLocks: ds.??? // Strage mapping + Not returned by connection.listDataset
            // CYlinders: ds.??? // Strage mapping + Not returned by connection.listDataset
            // TRacks: ds.??? // Strage mapping + Not returned by connection.listDataset
            // Directory: ds.??? // Strage mapping + Not returned by connection.listDataset
            // PRImary: ds.??? // Strage mapping + Not returned by connection.listDataset
            // SECondary: ds.??? // Strage mapping + Not returned by connection.listDataset
        };
    }
    /**
     * Allocate dataset.
     *
     * @param connection - zos-node-accessor connection
     * @param dsn - fully-qualified dataset name without quotes
     * @param option - Allocate option
     */
    public static async allocateDataSet(connection: any, dsn: string, option: IAllocateDataSetOption): Promise<void> {
        this.log.debug("Allocate data set '%s'", dsn);
        await connection.allocateDataset(dsn, option.dcb);
    }

    /**
     * Allocate a dataset using the properties from another dataset
     *
     * @param connection zos-doce-accessor connection
     * @param dsn fully-qualified dataset name without quotes to be allocated
     * @param like fully-qualified dataset name without quotes to be modeled after
     * @param options dataset allocation options to override the `like` dataset allocation details
     * @returns
     */
    public static async allocateLikeDataSet(connection: any, dsn: string, like: string, options?: IDataSetDetailedAllocationOptions): Promise<IDataSetDetailedAllocationOptions> {
        const newDs = await connection.listDataset(dsn) ?? [];
        if (newDs.length !== 0) {
            this.log.debug("Dataset %s already exists", dsn);
            return DataSetUtils.mapAllocationOptions(newDs);
        }

        this.log.debug("Allocate data set '%s' with similar attributes to '%s", dsn, like);
        const files = await connection.listDataset(like);

        this.log.debug("Found %d matching data sets", files.length);
        const filteredFiles: any[] = files?.map((file: any) => CoreUtils.addLowerCaseKeysToObject(file)) ?? [];
        if (filteredFiles.length === 0) {
            throw new ImperativeError({msg: "No datasets found: " + like});
        }
        const ds = filteredFiles.find((file: any) => file.dsname.toUpperCase() === like.toUpperCase());
        const option = { ...DataSetUtils.mapAllocationOptions(ds), ...(options ?? {})};

        this.log.debug("Allocation options to be used: %s", JSON.stringify(option));
        await connection.allocateDataset(dsn, option);
        return option;
    }

    /**
     * Helper method to split the dataset name from the member name
     *
     * @param name fully-qualified dataset name without quotes
     * @returns Object separating the dataset name and the member name
     */
    public static getDataSet(name: string): {dsn: string, member?: string} {
        const parts = name.replace(')', '').split('(');
        if (parts.length > 1) {
            return {
                dsn: parts[0],
                member: parts[1]
            };
        } else {
            return {
                dsn: name
            };
        }
    }

    /**
     * Copy Dataset
     *
     * @param connection zos-node-accessor connection
     * @param options All required options for copying a dataset
     * @param options.fromDsn fully-qualified SOURCE dataset name without quotes
     * @param options.toDsn fully-qualified TARGET dataset name without quotes
     * @param options.progress Optional: Task used to indicate the progress of this operation (not used yet)
     * @param options.replace Optional: Boolean used to force a dataset or member replacement
     */
    public static async copyDataSet(connection: any, options: ICopyDataSetOptions): Promise<void> {
        const fromDsn = DataSetUtils.getDataSet(options.fromDsn);
        this.log.debug("Source: %s", JSON.stringify(fromDsn));
        const toDsn = DataSetUtils.getDataSet(options.toDsn);
        this.log.debug("Target: %s", JSON.stringify(toDsn));

        if (!options.replace) {
            this.log.debug("Verify that the dataset (or member) does not already exist since we aren't allowed to replace it");
            const newDs = await connection.listDataset(toDsn.dsn) ?? [];
            if (newDs.length !== 0 && (await connection.listDataset(options.toDsn) ?? []).length !== 0) {
                throw new ImperativeError({msg: `Dataset ${options.toDsn} already exists`});
            }
        }

        const dataset = (await connection.listDataset(fromDsn.member ? fromDsn.dsn : options.fromDsn))?.map((f: any) => CoreUtils.addLowerCaseKeysToObject(f)) ?? [];
        if (dataset.length === 0) {
            throw new ImperativeError({msg: `The dataset "${fromDsn.dsn}" doesn't exist.`});
        }

        const content = dataset.find((file: any) => file.dsname.toUpperCase() === fromDsn.dsn.toUpperCase());
        this.log.debug("Select the transfer type based on the soruce dataset RECFM");
        let transferType = fromDsn.member ? TRANSFER_TYPE_ASCII : null;
        switch (content.recfm[0]) {
            case "V":
                transferType = TRANSFER_TYPE_BINARY_RDW;
                break;
            case "D":
                transferType = TRANSFER_TYPE_ASCII_RDW;
                break;
            default:
                transferType = transferType ?? TRANSFER_TYPE_BINARY;
                break;
        }

        this.log.debug("Download the contents of the source dataset: %s", JSON.stringify(options.fromDsn));
        const stream = await connection.getDataset("'"+options.fromDsn+"'", transferType, false);

        this.log.debug("Make sure the new dataset is allocated: %s", JSON.stringify(toDsn.dsn));
        const option = await DataSetUtils.allocateLikeDataSet(connection, toDsn.dsn, fromDsn.dsn, toDsn.member ? {dsorg: "PO"} : {dsorg: "PS"});

        this.log.debug("Upload the contents to the new dataset: %s", JSON.stringify(options.toDsn));
        await connection.uploadDataset(stream, "'"+options.toDsn+"'", transferType, option);
    }

    private static get log(): Logger {
        return Logger.getAppLogger();
    }
}
