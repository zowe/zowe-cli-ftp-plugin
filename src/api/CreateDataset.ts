/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { CreateDataSetTypeEnum, CreateDefaults, ICreateDataSetOptions, IZosFilesResponse, ZosFilesMessages } from "@zowe/cli";
import { ImperativeError, ImperativeExpect, TextUtils } from "@zowe/imperative";
import { JobUtils } from "./JobUtils";
import { CoreUtils } from "./CoreUtils";
import { AbstractTemplatedJCL } from "./AbstractTemplatedJCL";

export class CreateDataset extends AbstractTemplatedJCL {

    /**
     * Create a data set
     * @param connection                 - zos-node-accessor connection
     * @param  cmdType               - The type of data set we are going to create
     * @param dataSetName                          - the name of the data set to create
     * @param  [options={}] - additional options for the creation of the data set
     */
    public static async create(connection: any, cmdType: CreateDataSetTypeEnum,
        dataSetName: string,
        jobCardFile: string,
        options ?: Partial<ICreateDataSetOptions>):
        Promise<IZosFilesResponse> {
        let validCmdType = true;

        // Removes undefined properties
        let tempOptions = options != null ? JSON.parse(JSON.stringify(options)) : {};
        const secondarySpecified = tempOptions.secondary != null;

        // Required
        ImperativeExpect.toNotBeNullOrUndefined(cmdType, ZosFilesMessages.missingDatasetType.message);

        // Required
        ImperativeExpect.toNotBeNullOrUndefined(dataSetName, ZosFilesMessages.missingDatasetName.message);

        switch (cmdType) {
            case CreateDataSetTypeEnum.DATA_SET_PARTITIONED:
                tempOptions = {...CreateDefaults.DATA_SET.PARTITIONED, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_SEQUENTIAL:
                tempOptions = {...CreateDefaults.DATA_SET.SEQUENTIAL, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_BINARY:
                tempOptions = {...CreateDefaults.DATA_SET.BINARY, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_C:
                tempOptions = {...CreateDefaults.DATA_SET.C, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_CLASSIC:
                tempOptions = {...CreateDefaults.DATA_SET.CLASSIC, ...tempOptions};
                break;
            default:
                validCmdType = false;
                break;
        }

        if (!validCmdType) {
            throw new ImperativeError({msg: ZosFilesMessages.unsupportedDatasetType.message});
        } else {
            // Handle the size option
            if (tempOptions.size != null) {
                const tAlcunit = tempOptions.size.toString().match(/[a-zA-Z]+/g);
                if (tAlcunit != null) {
                    tempOptions.alcunit = tAlcunit.join("").toUpperCase();
                }

                const tPrimary = tempOptions.size.toString().match(/[0-9]+/g);
                if (tPrimary != null) {
                    tempOptions.primary = +(tPrimary.join(""));

                    if (!secondarySpecified) {
                        const TEN_PERCENT = 0.10;
                        tempOptions.secondary = Math.round(tempOptions.primary * TEN_PERCENT);
                    }
                }

                delete tempOptions.size;
            }

            let response = "";
            // Handle the print attributes option
            if (tempOptions.printAttributes != null) {
                if (tempOptions.printAttributes) {
                    delete tempOptions.printAttributes;
                    response = TextUtils.prettyJson(tempOptions);
                } else {
                    delete tempOptions.printAttributes;
                }
            }

            response = await new CreateDataset().createViaFTP(connection, dataSetName, tempOptions, jobCardFile);
            if (response.indexOf("RC=0000") >= 0) {
                return {
                    success: true,
                    commandResponse: response + "\n" + ZosFilesMessages.dataSetCreatedSuccessfully.message,
                }
            } else {
                return {
                    success: false,
                    commandResponse: response + "\nFailed to create data set",
                }
            }
        }
    }


    public DEFAULT_TEMPLATE = "//CREATE EXEC PGM=IEFBR14\n" +
        "//{{createDD}}";

    private async createViaFTP(connection: any, dataSetName: string,
        options: ICreateDataSetOptions,
        jobCardFile: string,
        overrideTemplateFile ?: string) {
        this.log.debug("Building jcl to create data set %s", dataSetName);

        let createDD: string = "NEWDATA DD ";
        const createDDValues = ["DSN=" + dataSetName.toUpperCase(), "DISP=(NEW,CATLG)"];
        const maxDDLineLength = 65;
        let currentLineLength = createDD.length;
        const continueLine = "//   ";

        // todo: make sure all options are used here
        if (options.alcunit) {
            let spaceParams = "SPACE=(";
            spaceParams += options.alcunit.toUpperCase() + ",(";
            spaceParams += options.primary;
            if (options.secondary != null) {
                spaceParams += "," + options.secondary;
            }
            spaceParams += "))";

            createDDValues.push(spaceParams);
            if (options.volser != null) {
                createDDValues.push("VOLSER=" + options.volser.toUpperCase());
            }
        }
        if (options.lrecl != null) {
            createDDValues.push("LRECL=" + options.lrecl);
        }
        if (options.recfm != null) {
            createDDValues.push("RECFM=" + options.recfm.toUpperCase());
        }
        if (options.blksize != null) {
            createDDValues.push("BLKSIZE=" + options.blksize);
        }
        if (options.unit != null) {
            createDDValues.push("UNIT=" + options.unit.toUpperCase());
        }

        if (options.dsorg != null) {
            createDDValues.push("DSORG=" + options.dsorg.toUpperCase());
        }

        /**
         * Build the DD definition for the job by looping through the values we've built
         */
        for (let i = 0; i < createDDValues.length; i++) {
            const value = createDDValues[i];
            if (value.length + currentLineLength > maxDDLineLength) {
                createDD += "\n" + continueLine;
                currentLineLength = continueLine.length;
            }

            createDD += value;
            if (i < createDDValues.length - 1) {
                // if this isn't the last value, add a comma to continue the line
                createDD += ",";
                currentLineLength += 1;
            }
            currentLineLength += value.length;
        }

        const jcl = this.getJcl(jobCardFile, {createDD}, overrideTemplateFile);
        this.log.debug("Creating data set %s via JCL:\n %s", dataSetName, jcl);
        const jobid = await connection.submitJCL(jcl);
        const jobWait = 500;
        await CoreUtils.sleep(jobWait); // have to wait for the job to show up in the list
        const jobDetails = await JobUtils.findJobByID(connection, jobid);
        return "Job details: " + JSON.stringify(jobDetails, null, 2);
    }

}
