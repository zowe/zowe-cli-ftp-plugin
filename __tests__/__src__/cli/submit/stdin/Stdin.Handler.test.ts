/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { CoreUtils } from "../../../../../src/api/CoreUtils";
import SubmitJobFromLocalFileHandler from "../../../../../src/cli/submit/stdin/Stdin.Handler";
import TestUtils from "../../TestUtils";

describe("Submit job from stdin handler", () => {

    it("should return correct message if the job is submit successfully.", async () => {
        const handler = new SubmitJobFromLocalFileHandler();
        const jobDetails = {
            jobname: "jobname1",
            jobid: "jobid1",
            spoolFiles: [
                {
                    name: "file1"
                },
                {
                    name: "file2"
                }
            ]
        };

        CoreUtils.readStdin = jest.fn().mockReturnValue(Promise.resolve(Buffer.from("sss")));
        const mockResponse = TestUtils.getMockResponse();
        const mockParams: any = {
            arguments: {
                file: "/u/user/jcl"
            },
            connection: {
                submitJCL: jest.fn().mockReturnValue(Promise.resolve("jobid1")),
                getJobStatus: jest.fn().mockReturnValue(Promise.resolve(jobDetails))
            },
            response: mockResponse
        };
        await handler.processFTP(mockParams);
        expect(mockResponse.data.setObj.mock.calls[0]).toMatchSnapshot();
        expect(mockResponse.format.output.mock.calls[0]).toMatchSnapshot();
    });

});
