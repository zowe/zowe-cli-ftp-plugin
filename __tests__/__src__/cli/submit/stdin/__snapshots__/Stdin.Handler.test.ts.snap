// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submit job from stdin handler should return correct message if the job is submit successfully. 1`] = `
Array [
  Object {
    "jobid": "jobid1",
    "jobname": "jobname1",
    "spoolFiles": Array [
      Object {
        "name": "file1",
      },
      Object {
        "name": "file2",
      },
    ],
  },
]
`;

exports[`Submit job from stdin handler should return correct message if the job is submit successfully. 2`] = `
Array [
  Object {
    "fields": Array [
      "jobid",
      "jobname",
      "owner",
      "status",
    ],
    "format": "object",
    "output": Object {
      "jobid": "jobid1",
      "jobname": "jobname1",
      "spoolFiles": Array [
        Object {
          "name": "file1",
        },
        Object {
          "name": "file2",
        },
      ],
    },
  },
]
`;
